chia-client/src/peer.rs

```diff
@@ -0,0 +1,219 @@
+use std::io::Cursor;
+use std::{collections::HashMap, sync::Arc};
+
+use chia_protocol::{
+    ChiaProtocolMessage, CoinStateUpdate, Handshake, Message, NewPeakWallet, NodeType,
+    ProtocolMessageTypes,
+};
+use chia_traits::Streamable;
+use futures_util::stream::SplitSink;
+use futures_util::{SinkExt, StreamExt};
+use tokio::sync::{broadcast, oneshot, Mutex};
+use tokio::{net::TcpStream, task::JoinHandle};
+use tokio_tungstenite::{MaybeTlsStream, WebSocketStream};
+
+use crate::utils::stream;
+use crate::{Error, Result};
+
+type WebSocket = WebSocketStream<MaybeTlsStream<TcpStream>>;
+type Requests = Arc<Mutex<HashMap<u16, oneshot::Sender<Message>>>>;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum PeerEvent {
+    CoinStateUpdate(CoinStateUpdate),
+    NewPeakWallet(NewPeakWallet),
+}
+
+pub struct Peer {
+    sink: Mutex<SplitSink<WebSocket, tungstenite::Message>>,
+    inbound_task: JoinHandle<()>,
+    event_receiver: broadcast::Receiver<PeerEvent>,
+    requests: Requests,
+    nonce: Mutex<u16>,
+}
+
+struct ResponseHandler {
+    requests: Requests,
+    message_id: u16,
+}
+
+impl ResponseHandler {
+    async fn new(
+        requests: Requests,
+        message_id: u16,
+    ) -> (ResponseHandler, oneshot::Receiver<Message>) {
+        let (sender, receiver) = oneshot::channel::<Message>();
+        requests.lock().await.insert(message_id, sender);
+
+        (
+            Self {
+                requests,
+                message_id,
+            },
+            receiver,
+        )
+    }
+}
+
+impl Drop for ResponseHandler {
+    fn drop(&mut self) {
+        let requests = Arc::clone(&self.requests);
+        let message_id = self.message_id;
+
+        tokio::spawn(async move {
+            requests.lock().await.remove(&message_id);
+        });
+    }
+}
+
+impl Peer {
+    pub fn new(ws: WebSocket) -> Self {
+        let (sink, mut stream) = ws.split();
+        let (event_sender, event_receiver) = broadcast::channel(32);
+
+        let requests = Requests::default();
+        let requests_clone = Arc::clone(&requests);
+
+        let inbound_task = tokio::spawn(async move {
+            while let Some(message) = stream.next().await {
+                if let Ok(message) = message {
+                    let bytes = message.into_data();
+                    let cursor = &mut Cursor::new(bytes.as_slice());
+
+                    // Parse the message.
+                    let Ok(message) = Message::parse(cursor) else {
+                        continue;
```

Comment:
this case might also warrant some logging, and maybe even closing the peer connection